syntax = "proto3";

package kvstore;

// Message types for ABD Protocol
message ABDReadRequest {
    string key = 1;
    int64 timestamp = 2;  // Client's timestamp
}

message ABDReadResponse {
    string value = 1;
    int64 timestamp = 2;  // Server's timestamp
    bool success = 3;
}

message ABDWriteRequest {
    string key = 1;
    string value = 2;
    int64 timestamp = 3;
}

message ABDWriteResponse {
    bool success = 1;
    int64 timestamp = 2;  // Server's timestamp
}

// Message types for Blocking Protocol
message BlockingLockRequest {
    string key = 1;
    int32 client_id = 2;
}

message BlockingLockResponse {
    bool granted = 1;
    int64 timestamp = 2;  // Server timestamp
}

message BlockingReadRequest {
    string key = 1;
    int32 client_id = 2;
}

message BlockingReadResponse {
    string value = 1;
    int64 timestamp = 2;
    bool success = 3;
}

message BlockingWriteRequest {
    string key = 1;
    string value = 2;
    int64 timestamp = 3;
    int32 client_id = 4;
}

message BlockingWriteResponse {
    bool success = 1;
    int64 timestamp = 2;
}

message BlockingUnlockRequest {
    string key = 1;
    int32 client_id = 2;
}

message BlockingUnlockResponse {
    bool success = 1;
}

// gRPC Service Definitions
service ABDService {
    rpc Read(ABDReadRequest) returns (ABDReadResponse);
    rpc Write(ABDWriteRequest) returns (ABDWriteResponse);
}

service BlockingService {
    rpc AcquireLock(BlockingLockRequest) returns (BlockingLockResponse);
    rpc Read(BlockingReadRequest) returns (BlockingReadResponse);
    rpc Write(BlockingWriteRequest) returns (BlockingWriteResponse);
    rpc ReleaseLock(BlockingUnlockRequest) returns (BlockingUnlockResponse);
}


